nixos/modules/services/web-servers/tomcat/default.nix

```diff
@@ -0,0 +1,480 @@
+            forceSSL = true;
+          };
+        '';
+        description = ''
+          List consisting of a virtual host name and a list of web applications to deploy on each virtual host
+        '';
+      };
+
+      logPerVirtualHost = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether to enable logging per virtual host.";
+      };
+
+      jdk = mkOption {
+        type = types.package;
+        default = pkgs.jdk;
+        defaultText = "pkgs.jdk";
+        description = "Which JDK to use.";
+      };
+
+    };
+  };
+
+  # implementation
+
+  config = mkIf cfg.enable {
+
+    assertions = [
+      {
+        assertion = all (hostOpts: with hostOpts; !(addSSL && onlySSL) && !(forceSSL && onlySSL) && !(addSSL && forceSSL)) vhosts;
+        message = ''
+          Options `services.httpd.virtualHosts.<name>.addSSL`,
+          `services.httpd.virtualHosts.<name>.onlySSL` and `services.httpd.virtualHosts.<name>.forceSSL`
+          are mutually exclusive.
+        '';
+      }
+      {
+        assertion = all (hostOpts: !(hostOpts.enableACME && hostOpts.useACMEHost != null)) vhosts;
+        message = ''
+          Options `services.httpd.virtualHosts.<name>.enableACME` and
+          `services.httpd.virtualHosts.<name>.useACMEHost` are mutually exclusive.
+        '';
+      }
+      {
+        assertion = hasAttr "defaultHost" cfg.engineAttrs && hasAttr cfg.engineAttrs.defaultHost cfg.virtualHosts;
+        message = ''
+          Please choose a default host for `tomcat` by setting `services.tomcat.engineAttrs.defaultHost` to one of: ${concatStringsSep ", " (builtins.attrNames cfg.virtualHosts)}
+        '';
+      }
+      {
+        assertion = cfg.connectors != [];
+        message = ''
+          Please add at least one tomcat connector to `services.tomcat.connectors`.
+        '';
+      }
+    ];
+
+    users.users = optionalAttrs (cfg.user == "tomcat") {
+      tomcat = {
+        group = cfg.group;
+        description = "Tomcat user";
+        uid = config.ids.uids.tomcat;
+        home = "/homeless-shelter";
+      };
+    };
+
+    users.groups = optionalAttrs (cfg.group == "tomcat") {
+      tomcat.gid = config.ids.gids.tomcat;
+    };
+
+    services.tomcat.javaOpts = [ "-Djava.awt.headless=true" ] ++ optional (any (connector: connector.protocol == "org.apache.coyote.http11.Http11AprProtocol") cfg.connectors) "-Djava.library.path=${pkgs.tomcat-native}/lib";
+
+    services.tomcat.serviceAttrs = mapAttrs (name: mkDefault) {
+      name = "Catalina";
+    };
+    services.tomcat.engineAttrs = mapAttrs (name: mkDefault) {
+      name = "Catalina";
+    } // optionalAttrs (builtins.length vhosts == 1) {
+      defaultHost = (head vhosts).hostName;
+    };
+
+    services.tomcat.extraConfig = ''
+      <Listener className="org.apache.catalina.startup.VersionLoggerListener" />
+      <Listener className="org.apache.catalina.core.AprLifecycleListener" SSLEngine="on" />
```

Comment:
actually, I had this specific line removed in my config:
```
              # disable tomcat native library (we don't want use tomcat as web server)
              sed -i '/<Listener className="org.apache.catalina.core.AprLifecycleListener" SSLEngine="on" \/>/d' $out/conf/server.xml
```
AFAIK the only reason to do this was to remove some warnings from logs.

Reply:
I'm going to look into this a bit more, but for starters I have made a `listeners` option of type `list` to declore `<Listener>` elements, allowing for much finer control if desired. As you mention a simple fix would be to conditionally add this depending on whether `tomcat-native` has been included.